library/include/rocwmma/internal/transforms_impl.hpp

```diff
@@ -527,6 +613,272 @@ namespace rocwmma
         }
     };
 
+    template <>
+    struct AosToSoa<16, 4>
+    {
+        constexpr static uint32_t VW      = 4;
+        constexpr static uint32_t VecSize = 4;
+
+        template <typename DataT>
+        ROCWMMA_DEVICE constexpr static inline auto exec(VecT<DataT, VecSize> const& v)
+        {
+            using PackUtil = PackUtil<DataT>;
+
+            // Step 1 : UnpackLoHi4
+            auto unpacked_data = unpackLoHi4(v);
+
+            // Step 2 : UnpackLoHi8
+            unpacked_data = unpackLoHi8(unpacked_data);
+
+            // Step 3 : Gather
+            unpacked_data = PackUtil::template paddedUnpack<4>(
+                Permute::Gather16<4, 0>::exec(PackUtil::paddedPack(unpacked_data)));
+
+            return unpacked_data;
+        }
+    };
+
+    template <>
+    struct AosToSoa<32, 4>
+    {
+        constexpr static uint32_t VW      = 4;
+        constexpr static uint32_t VecSize = 4;
+
+        template <typename DataT>
+        ROCWMMA_DEVICE constexpr static inline auto exec(VecT<DataT, VecSize> const& v)
+        {
+            using PackUtil = PackUtil<DataT>;
+
+            // Step 1 : UnpackLoHi8
+            auto unpacked_data = unpackLoHi8(v);
+
+            // Step 2 : UnpackLoHi16
+            auto lo       = PackUtil::paddedPack(extractEven(unpacked_data));
+            auto hi       = PackUtil::paddedPack(extractOdd(unpacked_data));
+            auto rot_hi   = Swizzle::RotateR32<16>::exec(hi);
+            hi            = Blend::Zip16::exec(rot_hi, lo);
+            lo            = Blend::Zip16::exec(lo, rot_hi);
+            unpacked_data = concat(PackUtil::template paddedUnpack<2u>(lo),
+                                   PackUtil::template paddedUnpack<2u>(hi));
+
+            // Step 3 : Gather
+            hi = Permute::Gather32<4, 16>::exec(PackUtil::paddedPack(extractHi(unpacked_data)));
+            lo = Permute::Gather32<4, 0>::exec(PackUtil::paddedPack(extractLo(unpacked_data)));
+            unpacked_data = concat(PackUtil::template paddedUnpack<2u>(lo),
+                                   PackUtil::template paddedUnpack<2u>(hi));
+
+            return unpacked_data;
+        }
+    };
+
+    template <>
+    struct AosToSoa<64, 4>
+    {
+        constexpr static uint32_t VW      = 4;
+        constexpr static uint32_t VecSize = 4;
+
+        template <typename DataT>
+        ROCWMMA_DEVICE constexpr static inline auto exec(VecT<DataT, VecSize> const& v)
+        {
+            auto unpacked_data = aos_soa_wave_b32_vw4(v);
```

Comment:
Recommend to put the wave implementation here (RDNA will be different anyway). Then from the 64+, call on this function

Reply:
done